test/iapetos/collector/ring_test.clj

```diff
@@ -11,90 +11,115 @@
-     wrap (gen/elements [ring/wrap-instrumentation ring/wrap-metrics])]
-    (let [registry   (registry-fn)
-          handler'   (wrap handler registry)
-          start-time (System/nanoTime)
-          response   (try
-                       (handler' request)
-                       (catch Throwable t
-                         ::error))
-          delta      (/ (- (System/nanoTime) start-time) 1e9)
-          labels     (merge labels labels')
-          ex-labels  (assoc labels' :exceptionClass "java.lang.Exception")
-          counter    (registry :http/requests-total labels)
-          histogram  (registry :http/request-latency-seconds labels)
-          ex-counter (registry :http/exceptions-total ex-labels)]
-      (if exception?
-        (and (= response ::error)
-             (= 0.0 (prometheus/value counter))
-             (= 0.0 (:count (prometheus/value histogram)))
-             (= 1.0 (prometheus/value ex-counter)))
-        (and (map? response)
-             (= 0.0 (prometheus/value ex-counter))
-             (< 0.0 (:sum (prometheus/value histogram)) delta)
-             (= 1.0 (prometheus/value counter)))))))
+   [registry-fn                                (g/registry-fn ring/initialize)
+    {:keys [handler async? exception? labels]} gen-handler
+    {labels' :labels, :as request}             gen-request
+    wrap                                       (gen/elements [ring/wrap-instrumentation ring/wrap-metrics])]
+   (let [registry   (registry-fn)
+         handler'   (wrap handler registry)
+         start-time (System/nanoTime)
+         response   (if async?
+                      (async-response handler' request)
+                      (sync-response handler' request))
+         delta      (/ (- (System/nanoTime) start-time) 1e9)
+         labels     (merge labels labels')
+         ex-labels  (assoc labels' :exceptionClass "java.lang.Exception")
+         counter    (registry :http/requests-total labels)
+         histogram  (registry :http/request-latency-seconds labels)
+         ex-counter (registry :http/exceptions-total ex-labels)]
+     (if exception?
+       (and (= response ::error)
+            (= 0.0 (prometheus/value counter))
+            (= 0.0 (:count (prometheus/value histogram)))
+            (= 1.0 (prometheus/value ex-counter)))
+       (and (map? response)
+            (= 0.0 (prometheus/value ex-counter))
+            (< 0.0 (:sum (prometheus/value histogram)) delta)
+            (= 1.0 (prometheus/value counter)))))))
 
 (defspec t-wrap-instrumentation-with-exception-status 10
   (prop/for-all
     [registry-fn                         (g/registry-fn ring/initialize)
-     {:keys [handler exception? labels]} gen-handler
+     {:keys [handler async? exception? labels]} gen-handler
      {labels' :labels, :as request}      gen-request
      wrap (gen/elements [ring/wrap-instrumentation ring/wrap-metrics])]
     (let [registry   (registry-fn)
-          ex-status  500 
+          ex-status  500
```

Comment:
By passing in this number as a boolean argument, it's expecting any errors to be caught and those requests counted. But I don't think it's as straightforward to catch an error in the async handler. It's not as simple as just wrapping the `(handler request response raise)` in the `(safe exception-status ...)` function they use for the synchronous version.